// @ngx-i18nsupport/ngx-i18nsupport-lib/esm2015/api/constants.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/** *
 * supported file formats
  @type {?} */
export const FORMAT_XLIFF12 = 'xlf';
/** @type {?} */
export const FORMAT_XLIFF20 = 'xlf2';
/** @type {?} */
export const FORMAT_XMB = 'xmb';
/** @type {?} */
export const FORMAT_XTB = 'xtb';
/** *
 * File types
 * (returned by fileType() method)
  @type {?} */
export const FILETYPE_XLIFF12 = 'XLIFF 1.2';
/** @type {?} */
export const FILETYPE_XLIFF20 = 'XLIFF 2.0';
/** @type {?} */
export const FILETYPE_XMB = 'XMB';
/** @type {?} */
export const FILETYPE_XTB = 'XTB';
/** *
 * State NEW.
 * Signals an untranslated unit.
  @type {?} */
export const STATE_NEW = 'new';
/** *
 * State TRANSLATED.
 * Signals a translated unit, that is not reviewed until now.
  @type {?} */
export const STATE_TRANSLATED = 'translated';
/** *
 * State FINAL.
 * Signals a translated unit, that is reviewed and ready for use.
  @type {?} */
export const STATE_FINAL = 'final';
/** *
 * Default format, contains placeholders, html markup.
  @type {?} */
export const NORMALIZATION_FORMAT_DEFAULT = 'default';
/** *
 * Format for usage in ngxtranslate messages.
 * Placeholder are in the form {{n}}, no html markup.
  @type {?} */
export const NORMALIZATION_FORMAT_NGXTRANSLATE = 'ngxtranslate';
